recipes/Text_to_SQL/Text_to_SQL.ipynb

```diff
@@ -0,0 +1,766 @@
+     ]
+    }
+   ],
+   "source": [
+    "question = \"Retrieve the list of users who posted tweets with a sentiment score below 0, grouped by their gender, along with the count of such tweets.\"\n",
+    "sql_query, result = get_answer_using_fewshot(db_schema, question)\n",
+    "print(f\"sql_query: {sql_query}\")\n",
+    "print(f\"result: {result}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**NOTE** - For the next question `Find the user(s) who posted the maximum number of tweets in a single day` the LLM couldn't generate the correct SQL query. It missed to include `per day` in the generated SQL query. This is fixed using the `Chain of Thought Reasoning` in the next section. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "sql_query: SELECT T.UserID, COUNT(*) AS total_tweets \n",
+      "FROM tweet T \n",
+      "JOIN user U ON T.UserID = U.UserID \n",
+      "GROUP BY T.UserID \n",
+      "ORDER BY total_tweets DESC \n",
+      "LIMIT 1; \n",
+      "result: [('U001', 3)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "question = \"Find the user(s) who posted the maximum number of tweets in a single day.\"\n",
+    "sql_query, result = get_answer_using_fewshot(db_schema, question)\n",
+    "print(f\"sql_query: {sql_query}\")\n",
+    "print(f\"result: {result}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### **Chain of Thought Reasoning**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Chain of Thought (CoT) prompting is used to improve the accuracy of complex reasoning tasks by guiding the model through a structured sequence of steps.  \n",
+    "\n",
+    "In this technique, we break down the task into a sequence of logical steps to help the language model understand the process required to generate an accurate SQL query."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "sql_query: SELECT T.UserID, T.Day, COUNT(*) AS total_tweets \n",
+      "FROM tweet T \n",
+      "GROUP BY T.UserID, T.Day \n",
+      "ORDER BY total_tweets DESC \n",
+      "LIMIT 1; \n",
+      "result: [('U001', 8, 2)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "question = \"\"\"Find the user(s) who posted the maximum number of tweets in a single day.\n",
+    "              First calculate tweets posted by each user on each day.\n",
+    "              Second sort the result in descending order of the number of tweets.\n",
+    "              Third findout the maximum number of tweets posted by a user in a single day.\n",
+    "              Finally select the user(s) who posted the maximum number of tweets in a single day.\"\"\"\n",
```

Comment:
@chakra4 can we try having the model generate its own plan, which is fed into the prompt in a second step? I'm thinking that someone writing an application using this technique will want to avoid having the end user write the request-specific chain-of-thought reasoning.

Reply:
@fayvor thank you for pointing that out. We will try the 2 step process tomorrow. I find one reference of such technique in https://arxiv.org/pdf/2304.11556

Reply:
Could that be a variant instead of a replacement?  I think there's pedagogical value in showing the plan hand-designed, before we get into generating plans.  I don't want to hold up this PR too long -- let's track that work in a separate PR (if not ticket).

Reply:
Yes, we can create a separate PR and work on it.

Reply:
Yes can raise a separate PR for that.